lightningd/payalgo.c

```diff
@@ -87,6 +89,120 @@ json_add_failures(struct json_result *r, char const *n,
 	json_array_end(r);
 }
 
+/* A point by which we can contact the payee.
+ * This represents a single 'r' field of the bolt11 invoice. */
+struct contactpoint {
+	/* Part of contact_points list of struct pay. */
+	struct list_node list;
+	/* Destination to search by getroute.
+	 * First node on route, or the actual destination if
+	 * subroute is empty array. */
+	struct pubkey to_find;
+	/* Total additional delay on subroute. */
+	u32 total_cltv_delta;
+	/* Route to use. Non-NULL tal_arr array. */
+	struct route_info *subroute;
+};
+
+/* Add the fees. */
+static void
+add_fee(u64 *amount, u32 base, u32 proportional)
+{
+	(*amount) += (*amount) * proportional / 1000000;
+	(*amount) += base;
+}
+
+/* Append a contactpoint route to the route found by getroute. */
+static void
+contactpoint_append(struct route_hop **route,
+		    const struct contactpoint *cp,
+		    /* Actual destination. */
+		    const struct pubkey *receiver_id,
+		    /* Actual value we will give to destination. */
+		    u64 try_msatoshi)
+{
+	size_t route_end;
+	size_t actual_route_size;
+	int i;
+	u64 amount;
+
+	if (tal_count(cp->subroute) == 0)
+		return;
+
+	route_end = tal_count(*route);
+	actual_route_size = route_end + tal_count(cp->subroute);
+
+	/* Initial route could start empty, if we are the
+	 * contactpoint.
+	 * If there is initial route, it should terminate
+	 * in the contact point to_find node.
+	 */
+	if (route_end != 0)
+		assert(structeq(&(*route)[route_end - 1].nodeid, &cp->to_find));
+
+	tal_resize(route, actual_route_size);
+
+	/* Load in reverse as we need to compute amounts in
+	 * reverse. */
+	/* FIXME: Unify amount-computing code here with amount-computing
+	 * code in gossipd/routing.c */
+	amount = try_msatoshi;
+	for (i = tal_count(cp->subroute) - 1; i >= 0; --i) {
+		(*route)[route_end + i].channel_id = cp->subroute[i].short_channel_id;
+		/* Get next hop, or dest if last on route. */
+		if (i != tal_count(cp->subroute) - 1)
+			(*route)[route_end + i].nodeid = cp->subroute[i + 1].pubkey;
+		else
+			(*route)[route_end + i].nodeid = *receiver_id;
+		(*route)[route_end + i].amount = amount;
+		(*route)[route_end + i].delay = (*route)[route_end + i - 1].delay - cp->subroute[i].cltv_expiry_delta;
```

Comment:
This line is wrong.  It assumes the previous delay is loaded.  But it is not loaded because we are loading route hops backward.  Instead, we should load the delay as + from next delay (or final target delay if this is the last hop).